libvast/native-plugins/rebuild.cpp

```diff
@@ -239,38 +239,42 @@ rebuilder(rebuilder_actor::stateful_pointer<rebuilder_state> self,
             // Adjust the counters, update the indicator, and move back
             // undersized transformed partitions to the list of remainig
             // partitions as desired.
-            VAST_ASSERT(!result.empty());
             bool needs_second_stage = false;
-            if (is_heterogeneous) {
-              self->state.num_heterogeneous -= 1;
-              if (result.size() > 1
-                  || result[0].events <= self->state.max_partition_size / 2) {
-                std::copy(result.begin(), result.end(),
-                          std::back_inserter(self->state.remaining_partitions));
-                needs_second_stage = true;
+            self->state.num_transformed += num_partitions;
+            self->state.num_results += result.size();
+            if (!result.empty()) {
+              if (is_heterogeneous) {
+                self->state.num_transformed -= num_partitions + 1;
+                self->state.num_results -= result.size();
+                self->state.num_heterogeneous -= 1;
```

Comment:
First unconditionally incrementing and then doing additional corrections in the branches is heavily confusing. Please rewrite this so that every branch does a single self-contained modification of these variables.